File: python/dsflix_sdk/modules/movies.py
from __future__ import annotations

from ..http import HttpClient


class MoviesModule:
    def __init__(self, http: HttpClient) -> None:
        self.http = http

    def get_popular(self, page: int | None = None, language: str | None = None, region: str | None = None) -> dict:
        return self.http.get("/api/v2/movies/popular", {"page": page, "language": language, "region": region})

    def get_trending(self, time_window: str = "day", language: str | None = None) -> dict:
        return self.http.get(f"/api/v2/trending/movie/{time_window}", {"language": language})

    def get_top_rated(self, page: int | None = None, language: str | None = None, region: str | None = None) -> dict:
        return self.http.get("/api/v2/movies/top-rated", {"page": page, "language": language, "region": region})

    def get_now_playing(self, page: int | None = None, language: str | None = None, region: str | None = None) -> dict:
        return self.http.get("/api/v2/movies/now-playing", {"page": page, "language": language, "region": region})

    def get_upcoming(self, page: int | None = None, language: str | None = None, region: str | None = None) -> dict:
        return self.http.get("/api/v2/movies/upcoming", {"page": page, "language": language, "region": region})

    def get_details(self, movie_id: int, language: str | None = None, append_to_response: str | None = None) -> dict:
        return self.http.get(f"/api/v2/movies/{movie_id}", {"language": language, "append_to_response": append_to_response})

    def get_cast(self, movie_id: int) -> dict:
        return self.http.get(f"/api/v2/movies/{movie_id}/cast")

    def search(self, query: str, page: int | None = None, language: str | None = None, year: int | None = None, include_adult: bool | None = None) -> dict:
        return self.http.get(
            "/api/v2/search/movie",
            {"query": query, "page": page, "language": language, "year": year, "include_adult": include_adult},
        )

    def download(self, media_id: int, quality: str | None = None) -> dict:
        return self.http.post("/api/v2/download", {"media_id": media_id, "quality": quality})

File: src/modules/movies.ts
// ─────────────────────────────────────────────────────────────
//  DSFlix SDK  –  src/modules/movies.ts
// ─────────────────────────────────────────────────────────────

import { HttpClient } from '../http';
import {
  PaginatedResult,
  Movie,
  MovieDetails,
  MovieListOptions,
  MovieDetailsOptions,
  MovieSearchOptions,
  Credits,
  DownloadParams,
  DownloadResponse,
} from '../types';

export class MoviesModule {
  constructor(private http: HttpClient) {}

  /** Get movies ordered by popularity */
  getPopular(options?: MovieListOptions): Promise<PaginatedResult<Movie>> {
    return this.http.get('/api/v2/movies/popular', options as Record<string, string | number | boolean | undefined>);
  }

  /** Get trending movies for 'day' or 'week' */
  getTrending(timeWindow: 'day' | 'week' = 'day', language?: string): Promise<PaginatedResult<Movie>> {
    return this.http.get(`/api/v2/trending/movie/${timeWindow}`, { language });
  }

  /** Get top rated movies */
  getTopRated(options?: MovieListOptions): Promise<PaginatedResult<Movie>> {
    return this.http.get('/api/v2/movies/top-rated', options as Record<string, string | number | boolean | undefined>);
  }

  /** Get movies currently in theatres */
  getNowPlaying(options?: MovieListOptions): Promise<PaginatedResult<Movie>> {
    return this.http.get('/api/v2/movies/now-playing', options as Record<string, string | number | boolean | undefined>);
  }

  /** Get upcoming movies */
  getUpcoming(options?: MovieListOptions): Promise<PaginatedResult<Movie>> {
    return this.http.get('/api/v2/movies/upcoming', options as Record<string, string | number | boolean | undefined>);
  }

  /** Get full details for a movie by ID */
  getDetails(movieId: number, options?: MovieDetailsOptions): Promise<MovieDetails> {
    return this.http.get(`/api/v2/movies/${movieId}`, options as Record<string, string | number | boolean | undefined>);
  }

  /** Get cast and crew for a movie */
  getCast(movieId: number): Promise<Credits> {
    return this.http.get(`/api/v2/movies/${movieId}/cast`);
  }

  /** Search movies by title */
  search(query: string, options?: MovieSearchOptions): Promise<PaginatedResult<Movie>> {
    return this.http.get('/api/v2/search/movie', { query, ...options } as Record<string, string | number | boolean | undefined>);
  }

  /** Request a download link for a movie */
  download(params: DownloadParams): Promise<DownloadResponse> {
    return this.http.post('/api/v2/download', params);
  }
}
